### PR TITLE
Fix the log duplication issue for --log-file

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:

when we use a `log-file`, all the items in l.file array point to the
same file. we are currently using a switch statement with fallthrough
which ends up logging the same line multiple times. So check if we are
using a `log-file` and just write the data once.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dims** [comment]
benchmark with "fix"

```
dims@bigbox:~/go/src/k8s.io/klog$ go test -benchmem -run=^$ k8s.io/klog -bench "^(BenchmarkLogs)$" -v -count=10
goos: linux
goarch: amd64
pkg: k8s.io/klog
BenchmarkLogs-20    	  200000	      8123 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	      7999 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	      7580 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	      8165 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	      7790 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	      8584 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	      7789 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	      7952 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	      7665 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	      8583 ns/op	     556 B/op	       6 allocs/op
PASS
ok  	k8s.io/klog	25.716s
```

benchmark without the "fix"
```
dims@bigbox:~/go/src/k8s.io/klog$ go test -benchmem -run=^$ k8s.io/klog -bench "^(BenchmarkLogs)$" -v -count=10
goos: linux
goarch: amd64
pkg: k8s.io/klog
BenchmarkLogs-20    	  100000	     10248 ns/op	     560 B/op	       6 allocs/op
BenchmarkLogs-20    	  100000	     10330 ns/op	     560 B/op	       6 allocs/op
BenchmarkLogs-20    	  100000	     10996 ns/op	     560 B/op	       6 allocs/op
BenchmarkLogs-20    	  100000	     10331 ns/op	     560 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	     10166 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  200000	     10168 ns/op	     556 B/op	       6 allocs/op
BenchmarkLogs-20    	  100000	     10920 ns/op	     560 B/op	       6 allocs/op
BenchmarkLogs-20    	  100000	     10079 ns/op	     560 B/op	       6 allocs/op
BenchmarkLogs-20    	  100000	     10913 ns/op	     560 B/op	       6 allocs/op
BenchmarkLogs-20    	  100000	     10162 ns/op	     560 B/op	       6 allocs/op
PASS
ok  	k8s.io/klog	24.772s
```

File: klog.go [hunk 1]
```diff
@@ -781,24 +781,36 @@ func (l *loggingT) output(s severity, buf *buffer, file string, line int, alsoTo
...
 			os.Stderr.Write(data)
 		}
-		if l.file[s] == nil {
-			if err := l.createFiles(s); err != nil {
-				os.Stderr.Write(data) // Make sure the message appears somewhere.
-				l.exit(err)
+
+		if logging.logFile != "" {
+			if l.file[infoLog] == nil {
+				if err := l.createFiles(s); err != nil {
```
Comment (neolit123) [inline klog.go]: should this be `l.createFiles(infoLog)`?
i don't know the details of the duplication issue, but does `logging.logFile` imply `infoLog` severity only?

Comment (dims) [reply klog.go]: Fixed @neolit123. added some comments to clarify what we are doing

**Comment by dims** [comment]
/assign @tallclair 

@yuwenma @tallclair can you please review since both of you worked on the previous attempt? (in https://github.com/kubernetes/kubernetes/pull/78466)

**Comment by yuwenma** [comment]
@dims  The PR lgtm. The only concern I have is that: Even if user provides --log-file flag, they should still be able to use --alsologtostderr flag to require standard output. This requires storing logs individually by severity level. Also, in such case users should be able to use --stderrthreshold to change the severity level of the standard output (e.g. if stderrthreshold is warning, all logs in warning and above warning will be printed in the stdout).

The above feature is no longer available after this PR goes in. Is this what we want? If so, can you add a docstring for alsologtostderr and stderrthreshold flags to explain this behavior?

File: klog.go [hunk 2]
```diff
@@ -781,24 +781,38 @@ func (l *loggingT) output(s severity, buf *buffer, file string, line int, alsoTo
...
 		}
-		if l.file[s] == nil {
-			if err := l.createFiles(s); err != nil {
-				os.Stderr.Write(data) // Make sure the message appears somewhere.
-				l.exit(err)
+
+		if logging.logFile != "" {
+			// Since we are using a single log file, all of the items in l.file array
+			// will point to the same file, so just use one of them to write data.
+			if l.file[infoLog] == nil {
```
Comment (tallclair) [inline klog.go]: nit: can we declare a constant to make this more explicit?
```
const singleFileLog = infoLog
```

**Comment by tallclair** [comment]
@yuwenma I don't understand your comment - it looks like the stderr output happens before these changes? Or were you commenting on an older version of this PR?

**Comment by yuwenma** [comment]
> @yuwenma I don't understand your comment - it looks like the stderr output happens before these changes? Or were you commenting on an older version of this PR?

yeah, i was referring to a old change.

/lgtm

**Pull request merged**